### PR TITLE
table: Speed up the add index operation

### PR DESCRIPTION
**Pull request by zimulala**
Using 4 TiKV, 1 PD, 2 TiDB , the servers of TiKV and TiDB are not on a computer.
The number of records in this table is 100,000,000, and the table's structure is as follows:
CREATE TABLE tt (
a int(11) DEFAULT NULL,
b varchar(256) DEFAULT NULL,
c varchar(256) DEFAULT NULL,
d int(11) DEFAULT NULL,
e int(11) DEFAULT NULL,
f int(11) DEFAULT NULL)ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin

tidb-before :  adding an index f takes time 35m33.008533773s
tidb-current: adding an index d takes time 35m1.221069105s

### CODE REVIEW

**Comment by shenli** [comment]
Seems there is no improvement?

File: table/tables/index.go [hunk 1]
```diff
@@ -97,6 +97,8 @@ type index struct {
 	tblInfo *model.TableInfo
 	idxInfo *model.IndexInfo
 	prefix  kv.Key
+
+	key []byte // It's used reduce the number of new slice when multiple index keys are created.
```
Comment (coocood) [inline table/tables/index.go]: This may be accessed by multiple goroutines.

Comment (winkyao) [reply table/tables/index.go]: Maybe you can use 2d array, get key array by keys[goroutine unique id]

Comment (coocood) [reply table/tables/index.go]: We can add an argument for the caller to pass in the key buffer.

**Comment by zimulala** [comment]
PTAL @coocood @winkyao

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
How about use `colMap map[int64]*types.FieldType` directly?

**Comment by zimulala** [comment]
I need used this structure in #4437

**Comment by zimulala** [comment]
@coocood I closed #4437, so I address your comments.

**Comment by shenli** [comment]
@winkyao @breeswish PTAL

**Pull request merged**